### ISSUE TITLE
`struct_excessive_bools` false positive for booleans in CLI arguments

### ISSUE THREAD

**Issue by oxalica** [rust-lang/rust-clippy#10923]
### Summary

The lint triggers even when we truly want to boolean productions, where booleans are not exclusive and could not be simplified with enums.

### Lint Name

struct_excessive_bools

### Reproducer

This is a common struct for CLI arguments:

```rust
struct App {
    force: bool,
    no_clobber: bool,
    interactive: bool,
    verbose: bool,
    // files, ...
}
```

It emits:

```
warning: more than 3 bools in a struct
 --> src/lib.rs:1:1
  |
1 | / pub struct App {
2 | |     pub force: bool,
3 | |     pub no_clobber: bool,
4 | |     pub interactive: bool,
5 | |     pub verbose: bool,
6 | |     // files, ...
7 | | }
  | |_^
  |
  = help: consider using a state machine or refactoring bools into two-variant enums
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#struct_excessive_bools
  = note: requested on the command line with `-W clippy::struct-excessive-bools`
```

The help does not work in this case, since we cannot replace it with enum in a simple way. We actually want to accept and handle all $2^4$ possibilities.

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-unknown-linux-gnu
release: 1.70.0
LLVM version: 16.0.2
```

**Comment by Centri3** [comment]
Not really a false positive, this is exactly why it's allow-by-default :wink: Finding whether booleans are exclusive or not is tricky, as that would require global analysis, outside the scope of Clippy. imo you should just allow the lint here

This should probably be moved to `restriction`, though; `pedantic` is a bit of a weird choice imo

**Comment by Centri3** [comment]
It's worth mentioning though that you can refactor this to use enums, though (like, `Force::Force` or `Force::Try`). So this lint is still correct. Just not with clap's derive API.

**Comment by oxalica** [comment]
> Not really a false positive, this is exactly why it's allow-by-default

The reasoning of why we warn this (disjoint states) does not apply to the code, so I think this is not expected to be emitted here, thus false-positive. "pedantic" group is aggressive but should still be generally reasonable. There are many projects enables pedantic groups by default like clippy itself. At least I want to mark this as an unresolved issue for that lint.

> you can refactor this to use enums, though (like, `Force::Force` or `Force::Try`)

This is about [algebraic blindness](https://github.com/quchen/articles/blob/1f33a89776f11e03873ba4fc5a81b90bd50eb342/algebraic-blindness.md) which applies to any boolean arguments, not necessary here. Every boolean can be more readable by making their variants named as enum. That may deserve a separated lint but is orthogonal to this one. Here, we already have field names.  `if args.force { .. }` is pretty clear.

> This should probably be moved to `restriction`, though; `pedantic` is a bit of a weird choice imo

That would suppress the warning but make it less discoverable. Is there local condition we can check before emitting this lint?

**Comment by Centri3** [comment]
> "pedantic" group is aggressive but should still be generally reasonable

This is reasonable, it's telling you exactly what you should do - refactor to use two-variant enums.

> That may deserve a separated lint but is orthogonal to this one.

No, both are the same lint. This lint is just destined to lint all cases of many `bool`s - whether it be a state machine where they're all mutually exclusive or not.

Ideally, they would be different lints but checking whether they're mutually exclusive requires global analysis which clippy will not do.

> `if args.force { .. }` is pretty clear.

Indeed, but what if you passed `force` to a function that expected `no_clobber`? Or checked `verbose` instead of `force`? It's not just trying to be more readable - It's trying to also preventing bugs.

(Though I do agree, `args.force` is very clean. It's unfortunate but very understandable you cannot implicitly convert a two-variant to a bool.)

> Is there local condition we can check before emitting this lint?

No. As stated before, clippy won't do global analysis.

Maybe for clap in particular, you can check whether all members of a struct are exclusive or `group(multiple = false)`, but clippy also doesn't often have lints added for external crates and it doesn't really solve this issue particularly well.

**Comment by oxalica** [comment]
> > `if args.force { .. }` is pretty clear.
> 
> Indeed, but what if you passed `force` to a function that expected `no_clobber`? Or checked `verbose` instead of `force`? It's not just trying to be more readable - It's trying to also preventing bugs.

I'd argue that in this situation. we should blame the function accepting one or more booleans, not the struct. As mentioned in the link I given above, function arguments are anonymous when called `func(true /* what's true? */)`, while struct fields are always named, both for reading and writing. That's why I think they are orthogonal and multiple *named* struct fields with the same type (no matter it's bool or not) themselves do not cause more confusion. When calling functions with boolean arguments, besides bool structs fields, bool literals, bool variables, or anything else could also trigger that mismatching bug you said.

If we do want to prevent the [algebraic blindness](https://github.com/quchen/articles/blob/1f33a89776f11e03873ba4fc5a81b90bd50eb342/algebraic-blindness.md) issue, we'd better limit the boolean usages in anonymous/positional places, like tuple, tuple structs and function arguments, not just named struct fields -- that's the least problematic place.
